BitFaster.Caching/Buffers/MpscBoundedBuffer.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BitFaster.Caching.Buffers
+{
+    /// <summary>
+    /// Provides a multi-producer, single-consumer thread-safe ring buffer. When the buffer is full,
+    /// TryAdd fails and returns false. When the buffer is empty, TryTake fails and returns false.
+    /// </summary>
+    /// <remarks>
+    /// Based on BoundedBuffer by Ben Manes.
+    /// https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedBuffer.java
+    /// </remarks>
+    public class MpscBoundedBuffer<T> where T : class
+    {
+        private T[] buffer;
+        private readonly int mask;
+        private PaddedHeadAndTail headAndTail; // mutable struct, don't mark readonly
+
+        public MpscBoundedBuffer(int boundedLength)
+        {
+            if (boundedLength < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(boundedLength));
+            }
+
+            // must be power of 2 to use & slotsMask instead of %
+            boundedLength = BitOps.CeilingPowerOfTwo(boundedLength);
+
+            buffer = new T[boundedLength];
+            mask = boundedLength - 1;
+        }
+
+        public int Capacity => buffer.Length;
+
+        public int Count
+        {
+            get
+            {
+                var spinner = new SpinWait();
+                while (true)
+                {
+                    var headNow = Volatile.Read(ref headAndTail.Head);
+                    var tailNow = Volatile.Read(ref headAndTail.Tail);
+
+                    if (headNow == Volatile.Read(ref headAndTail.Head) &&
+                        tailNow == Volatile.Read(ref headAndTail.Tail))
+                    {
+                        return GetCount(headNow, tailNow);
+                    }
+
+                    spinner.SpinOnce();
+                }
+            }
+        }
+
+        private int GetCount(int head, int tail)
+        {
+            if (head != tail)
+            {
+                head &= mask;
+                tail &= mask;
+
+                return head < tail ? tail - head : buffer.Length - head + tail;
+            }
+            return 0;
+        }
+
+        // thread safe
+        public BufferStatus TryAdd(T item)
+        {
+            int head = this.headAndTail.Head;
+            int tail = Volatile.Read(ref this.headAndTail.Tail);
```

Comment:
fwiw, I did a volatile read for a happens-before edge (acquire / load-load) so that the other plain read could piggyback on it for free. You have it before so you are more likely to get a stale read. Probably doesn’t matter much though.

Reply:
Piggybacking is some next level thinking.

I hadn't spotted that readCounter is declared volatile in your code (which was careless because I thought I was paying attention). 

I don't know if .NET has the same visibility guarantee as Java - whether non volatile fields are also refreshed. I need to think hard about this. The Volatile.Read static method actually inserts a full fence afaik, so probably VolatileRead(head) first would do the trick here.

Reply:
I switched both head and tail to volatile. So now there are 2 acquire reads, instead of 1 full fence and 1 plain (since I no longer call Volatile.Read or Volatile.Write which each internally call Thread.MemoryBarrier). Eliminating the full fences makes it about 7.5% faster in the microbenchmark. 

In .NET there is no built-in way to do a non-volatile read of a volatile field, or a volatile read of a non-volatile field without inserting a full fence. The APIs are not as good as Java or C++ which are much more explicit and give all options.

The only workaround I can think of would be to make a union by placing 2 int fields in the PaddedHeadAndTail struct with the same offset and make only 1 volatile. I feel like that could work, but this is all pretty esoteric stuff so I'm not certain.

The other problem is that head and tail are integers, and I think this code will break when they wrap.

Reply:
I think it is probably fine as volatile reads. I suppose technically one could claim the map lookup provided an early enough acquire fence, as this is called only on a hit. I play with memory barriers to learn the hardware, but I don't think it measurably impacts performance. My comment was more about possible confusion by swapping the order.

I don't recall why I used a `long`, maybe out of habit? It shouldn't break due to an overflow since it uses the delta as a guard and a mask for the index. It's a good test case that I should to add, though.

Reply:
You're right. I thought the size computation would break, but actually I think it is correct because when tail has wrapped to -2147483647 subtracting a large head number will itself wrap the result back (overflow is unchecked by default).

Reply:
fwiw, I borrowed the striping logic for buffers from Java's concurrent counters (LongAdder). Those counters are what I use if metrics are enabled, e.g. the hit counter. Since the read buffer is lossy it didn't seem appropriate to rely on maintenance. The lazily initialized, scalable counters serve the same purpose of trying to avoid memory waste and trying to give each cpu its own counter to minimize CAS conflicts. This of course has to retry until incremented and the counters summed for the total value.

I raise this because if you end up writing a similarly striped class for the read buffers, you are 90% there for scalable counters. While enabling stats will degrade somewhat, it won't be nearly as substantial so as to make a difference for end users. That's a useful utility class in its own right, so I'm surprised .NET doesn't have one already and your own could be a useful standalone utility.